packages/apify/src/internals/charging.ts

```diff
@@ -0,0 +1,281 @@
+        const chargedCount = Math.min(count, this.calculateMaxEventChargeCountWithinLimit(eventName));
+
+        if (chargedCount === 0) {
+            return {
+                eventChargeLimitReached: true,
+                chargedCount: 0,
+                chargeableWithinLimit: calculateChargeableWithinLimit(),
+            };
+        }
+
+        const pricingInfo = this.pricingInfo[eventName] ?? {
+            price: 0,
+            title: 'Unknown event',
+        };
+
+        this.chargingState[eventName] ??= { chargeCount: 0, totalChargedAmount: 0 };
+        this.chargingState[eventName].chargeCount += chargedCount;
+        this.chargingState[eventName].totalChargedAmount += chargedCount * pricingInfo.price;
+
+        /* END OF CRITICAL SECTION */
+
+        if (this.isAtHome) {
+            if (this.pricingInfo[eventName] !== undefined) {
+                await this.apifyClient.run(this.actorRunId!).charge({ eventName, count: chargedCount });
+            } else {
+                log.warning(`Attempting to charge for an unknown event '${eventName}'`);
+            }
+        }
+
+        const timestamp = new Date().toISOString();
+
+        for (let i = 0; i < chargedCount; i++) {
+            await this.chargingLogDataset.pushData({
+                eventName,
+                eventTitle: pricingInfo.title,
+                eventPriceUsd: pricingInfo.price,
+                timestamp,
+            });
+        }
+
+        if (chargedCount < count) {
+            log.info(`Charging ${count} instances of ${eventName} event would exceed maxTotalChargeUsd - only ${chargedCount} events were charged`);
+        }
+
+        return {
+            eventChargeLimitReached: this.calculateMaxEventChargeCountWithinLimit(eventName) <= 0,
+            chargedCount,
+            chargeableWithinLimit: calculateChargeableWithinLimit(),
+        };
+    }
+
+    /**
+     * Get the maximum amount of money that the Actor is allowed to charge.
+     */
+    getChargedEventCount(eventName: string): number {
+        if (this.chargingState === undefined) {
+            throw new Error('ChargingManager is not initialized');
+        }
+
+        return this.chargingState[eventName]?.chargeCount ?? 0;
+    }
+
+    /**
+     * Get the number of events with given name that the Actor has charged for so far.
+     */
+    getMaxTotalChargeUsd(): number {
+        if (this.chargingState === undefined) {
+            throw new Error('ChargingManager is not initialized');
+        }
+
+        return this.maxTotalChargeUsd;
+    }
+
+    private calculateTotalChargedAmount(): number {
+        if (this.chargingState === undefined) {
+            throw new Error('ChargingManager is not initialized');
+        }
+
+        const result = Object.values(this.chargingState).map(({ totalChargedAmount }) => totalChargedAmount).reduce((sum, inc) => sum + inc, 0);
+
+        // Keeping float precision issues at bay
+        return Number(result.toFixed(6));
```

Comment:
Q: is this actually helpful? Wouldn't any float precision issues actually happen after the 6th decimal digit? I guess yes, in which case you are just rounding here and introducing potentially bigger discrepancies than the possible native floating point issues.

For context, I think 6 decimals should be more than enough, but I don't t think this is actually necessary here.

Reply:
This was taken verbatim from the PoC made by @metalwarrior665 - I guess he can provide a better answer than me.

Reply:
And I took it verbatim from someone else :D Please don't take me as expert in floats and do your own research. But this at least fixes the issue with the sum
```
(0.1 + 0.2) // => // 0.30000000000000004
```

```
Number((0.1 + 0.2).toFixed(6))  // => 0.3
```

This then makes the problem when calculating how many events we can still do. As this exact problem happened on the platform recently, we should have unified solution for it